README.md
# Companion V2
**Afficheur déporté pour MSunPV**

![Screenshot](img/IMG_6869.jpg) ![Screenshot](img/IMG_6871.jpg)

Le Companion est un écran déporté pour suivre la production de panneaux photovoltaiques, la consommation de votre domicile et la recharge solaire de votre cumulus, ces données étant récupérées depuis un routeur solaire M'SunPV.

Cet afficheur est conçu pour être utilisé avec le LILYGO T-Display S3 et certaines bibliothèques sont nécessaires pour un fonctionnement correct sous Arduino IDE. Tous les détails sont dans cette vidéo https://youtu.be/8KDCmyMWrUk ainsi que dans le document pdf de ce repository.


Par ailleurs une clef API gratuite openweathermap.org est nécessaire (voir sur leur site). Créez un compte gratuit pour ensuite obtenir une clef API directement utilisable sur votre Companion. Les problèmes de réception des données OpenWeather ont été corrigés.


**Fonctions de l'afficheur**

Cette version permer l'affichage de :
* La production des panneaux photovoltaiques installés
* La consommation instantanée de votre domicile
* La consommation utilisée vers le cumulus et sa température de chauffe actuelle
* Les cumuls journaliers de la production, de la consommation, de la chauffe cumulus et de la ré-injection
* Un bouton assistant vous indiquant la possibilité de lancer une machine énergivore
* La date et l’heure actuelle (avec gestion automatique des heures été/hiver)
* La météo et la température extérieure actuelle
* L'heure du lever et coucher de soleil
* Mise en veille programmable et réactivation automatique
* En cas de perte de signal, récupération automatique
* Divers voyants indiquant la mise en route d’un chauffage électrique ou le point de givre
* Puissance du réseau wifi affichée, intéressant pour positionner l'afficheur alimenté par piles
* Fonction serveur web avec possibilité de lire les données depuis un navigateur
* Accès au serveur web avec l'invite http://companion dans un navigateur

![Screenshot](img/affiche.jpeg) 

Le rafraichissement des données photovoltaiques se fait toutes les 15 secondes, celui de la météo toutes les 15 minutes.

Le bouton du T-Diplay situé en bas, si vous avez branché le module par la gauche affiche les cumuls de la consommation, de la production des panneaux, de la réinjection, et de la charge du cumulus.

Le bouton haut permet de régler la luminosité de l'écran avec effet de va-et-vient et rendu graphique. La charge de la batterie éventuellement installée est visualisée également. 
En cas de problème, le bouton reset situé sur le haut du boitier permet de relancer le programme.

**MSunPV**

C'est un routeur solaire permettant d'utiliser le courant normalement réinjecté dans le réseau et de l'utiliser pour recharger un cumulus, lancer une climatisation, faire fonctionner un radiateur, un moteur ou une pompe par exemple. Tous les détails sont sur le site https://ard-tek.com

![Screenshot](img/SAM_0251_640.JPG)


Changelog
07/05/2023
Version 2.53 & 2.54
Corrections apportées pour l'affichage du voyant assistant

25/04/2023
Version 2.51
Amélioration du bargraph consommation
Visualisation consommation/nulle/injection
Affichage de la température chauffe cumulus avec décimale
Correction de bugs.

23/04/2023
Version 2.50
Ajout de la puissance su signal wifi sur l'écran principal (Attention le fichier images.h a été modifié)
Inversion du sens de la jauge éclairage (dans le même sens que les autres)
Voyant bleu remplacé par un voyant orange
Activation du mDNS responder: encore idée de Bellule. ll suffit de taper dans l'invite du navigateur web
cette simple commande http://companion pour y accéder

13/04/2023
Version 2.45
Correction de l'affichage de la date avant le changement effectif
Personnalisation de l'intensité d'affichage

11/04/2023
Version 2.40
Reconnexion vraiment automtique en cas de perte de signal
Serveur web amélioré

05/04/2023
Version 2.37
Correction bug empechant sorti de veille
Corrections typo sur serveur web

03/04/2023
Version 2.36
Nécessite l'installation de la bibliothèque OneButton
En mode veille, l'appui sur éclairage affiche les données pendant 5 sec.
Un double clic annule le mode veille.

02/04/2023
Version 2.34
Optimisation du serveur web (idée de Bellule)
Affichage annuel du poisson d'avril ou pas

01/04/2023
Version 2.33
Reconnexion en cas de perte de signal
Affichage de la perte en dB sur la page de garde

29/03/2023
Version 2.32
Censé se reconnecter après perte de signal. Dites moi.

28/03/2023
Version 2.31
Correction d'un bug, l'affichage des horaires lever et coucher de soleil ne tenait pas compte du décalage heure été/hiver

14/03/2023
Version 2.30
Ajout de la fonction serveur web (conception et idée de Bellule)
Vous avez la possibilité de lire les infos depuis un navigateur web connecté à l'adresse du Companion

18/02/2023
Version 2.21 -  Veille désactivable

16/02/2023
Version 2.20 -  Correction du relevé météo qui n'était accessible qu'avec des clés API nécessitant un abonnement. Les clés gratuites fonctionnent 
sans problème désormais. 
Ajout de la fonction veille lorsque la production des panneaux est nulle la nuit.

11/02/2023
Correction d'un affichage erroné du bargraph lorsque la consommation est égale à 0.

08/02/2023
Calcul du lever et coucher de soleil optimisés.
Indication de la météo, de la température extérieure, du point de givre.
Changement de police pour l'affichage de la page cumuls.
Correction de divers bugs.
Amélioration des commentaires dans le programme principal.

01/02/2023
Affichage des horaires de lever et coucher de soleil lorsque les panneaux sont hors service
Affichage correct de la date en français
Correction de quelques bugs
Attention, il est nécessaire d'installer la bibliothèque SolarCalculator par jpb10 que vous recherchez dans les bibliothèques Arduino.
Les variables personnelles sont désormais dans le fichier perso.h

18/01/2023
La température de chauffe du cumulus est visible, un anneau lumineux varie en fonction de sa température.
Il est nécessaire de déclarer la sonde en début du programme pour que l'affichage se fasse.

15/01/2023
Affichage correct des jauges en tenant compte de la puissance des panneaux installés et de la consommation max d'un cumulus.
Ces données sont à saisir après les données nécessaires pour se connecter au wifi de votre domicile

12/01/2023
Modification permettant d'avoir un affichage correct si la production des panneaux est déclarée négative
(Ajout de la ligne 373)



perso.h
/******************************************************************************
**************************  Données personnelles  *****************************
****************************** (version 2.30) *********************************
***    Mettez ici toutes vos valeurs personnelles pour le fonctionnement    ***
***             correct de l'afficheur externe Companion                    ***
*******************************************************************************
***     >>>>>  Les valeurs correspondantes à vos branchements               ***
***            et sondes sont à modifier éventuellement                     ***
***            dans le fichier companion.ino                                ***
***            aux lignes 627 et suivantes pour les index                   ***
***            et 674 et suivantes pour les cumuls                          ***
******************************************************************************/

// Vos codes accès au wifi. Remplacez les * par vos valeurs
const char* ssid     = "******";
const char* password = "******";

// Adresse IP du serveur local (ici MSunPV). Il se peut suivant votre réseau
// que vous ayiez besion de remplacer toute la chaine, respectez . entre les chiffres
char serveur[] = "192.168.1.**";

// Boitier horizontal prise à gauche, pour prise à droite mettez rotation = 1
int rotation = 3;

// Les panneaux et le cumulus peuvent avoir une mini conso de veille
// Mettez ici celle que vous estimez pour votre installation
int residuel = 10; // en Watt

// Affichage en nombres entiers
// Affiche aussi les valeurs en kw quand nécessaire
bool nbrentier = true;  // ou false pour chiffres avec virgule

// Eclairage intermédiaire au lancement et en sortie de veille
// Modifiable mais mettez IMPERATIVEMENT des multiples de 50
// (les valeurs min/max sont 0/250)
int dim_choisie = 100;      // Eclairage choisi au lancement

// Modifiez les lignes suivantes en fonction de votre équipement
int puissance = 5000; // production max en watt
int cumulus = 3000; // puissance cumulus en watt

// Si l'alimentation se fait par batterie (true pour oui et false pour non)
bool lipo = false;

// Chauffage par radiateurs électriques, mettez false sinon
bool chauffageElectr = true;

// Affichage de température si vous avez installé une sonde sur le cumulus
// Mettez alors à true et vérifiez ligne 587 dans companion.ino
bool sonde = false;

// Mise en veille affichage écran quand PV=0 (idée de Defaliz)
bool veille = false;

// Localisation de votre ville et décalage horaire pour lever/coucher soleil
String latitude =  "44.8378"; // 90.0000 to -90.0000 negative for Southern hemisphere
String longitude = "-0.594"; // 180.000 to -180.000 negative for West

// Données pour openweathermap.org ( API key et ville)
String api_key = "**************************";
String units = "metric";  // ou "imperial"
String language = "fr";

// Clin d'oeil pour le 1er Avril
// Fait apparaitre un poisson à la place de l'icone météo ce jour là
bool poisson = false;

//*****************************************************************************
//*******************    Fin des données personnelles    **********************
//*****************************************************************************


NTP_Time.h
//====================================================================================
//                                  Libraries
//====================================================================================

// Time library:
// https://github.com/PaulStoffregen/Time
#include <Time.h>

// Time zone correction library:
// https://github.com/JChristensen/Timezone
#include <Timezone.h>
#include <WiFi.h>
#include <WiFiUdp.h>

// A UDP instance to let us send and receive packets over UDP
WiFiUDP udp;

//====================================================================================
//                                  Settings
//====================================================================================
// Try to use pool url instead so the server IP address is looked up from those available
// (use a pool server in your own country to improve response time and reliability)
//const char* ntpServerName = "time.nist.gov";
//const char* ntpServerName = "pool.ntp.org";
const char* ntpServerName = "time.google.com";

// Try not to use hard-coded IP addresses which might change, you can if you want though...
//IPAddress timeServerIP(129, 6, 15, 30);   // time-c.nist.gov NTP server
//IPAddress timeServerIP(24, 56, 178, 140); // wwv.nist.gov NTP server
IPAddress timeServerIP;                     // Use server pool

// Example time zone and DST rules, see Timezone library documents to see how
// to add more time zones https://github.com/JChristensen/Timezone

// Zone reference "UK" United Kingdom (London, Belfast)
TimeChangeRule BST = {"BST", Last, Sun, Mar, 1, 60};        //British Summer (Daylight saving) Time
TimeChangeRule GMT = {"GMT", Last, Sun, Oct, 2, 0};         //Standard Time
Timezone UK(BST, GMT);

// Zone reference "euCET" Central European Time (Frankfurt, Paris)
TimeChangeRule CEST = {"CEST", Last, Sun, Mar, 2, 120};     //Central European Summer Time
TimeChangeRule  CET = {"CET ", Last, Sun, Oct, 3, 60};      //Central European Standard Time
Timezone euCET(CEST, CET);

// Zone reference "ausET" Australia Eastern Time Zone (Sydney, Melbourne)
TimeChangeRule aEDT = {"AEDT", First, Sun, Oct, 2, 660};    //UTC + 11 hours
TimeChangeRule aEST = {"AEST", First, Sun, Apr, 3, 600};    //UTC + 10 hours
Timezone ausET(aEDT, aEST);

// Zone reference "usET US Eastern Time Zone (New York, Detroit)
TimeChangeRule usEDT = {"EDT", Second, Sun, Mar, 2, -240};  //Eastern Daylight Time = UTC - 4 hours
TimeChangeRule usEST = {"EST", First, Sun, Nov, 2, -300};   //Eastern Standard Time = UTC - 5 hours
Timezone usET(usEDT, usEST);

// Zone reference "usCT" US Central Time Zone (Chicago, Houston)
TimeChangeRule usCDT = {"CDT", Second, dowSunday, Mar, 2, -300};
TimeChangeRule usCST = {"CST", First, dowSunday, Nov, 2, -360};
Timezone usCT(usCDT, usCST);

// Zone reference "usMT" US Mountain Time Zone (Denver, Salt Lake City)
TimeChangeRule usMDT = {"MDT", Second, dowSunday, Mar, 2, -360};
TimeChangeRule usMST = {"MST", First, dowSunday, Nov, 2, -420};
Timezone usMT(usMDT, usMST);

// Zone reference "usAZ" Arizona is US Mountain Time Zone but does not use DST
Timezone usAZ(usMST, usMST);

// Zone reference "usPT" US Pacific Time Zone (Las Vegas, Los Angeles)
TimeChangeRule usPDT = {"PDT", Second, dowSunday, Mar, 2, -420};
TimeChangeRule usPST = {"PST", First, dowSunday, Nov, 2, -480};
Timezone usPT(usPDT, usPST);


//====================================================================================
//                                  Variables
//====================================================================================
TimeChangeRule *tz1_Code;   // Pointer to the time change rule, use to get the TZ abbrev, e.g. "GMT"

time_t utc = 0;

bool timeValid = false;

unsigned int localPort = 2390;      // local port to listen for UDP packets

const int NTP_PACKET_SIZE = 48; // NTP time stamp is in the first 48 bytes of the message

byte packetBuffer[ NTP_PACKET_SIZE ]; //buffer to hold incoming and outgoing packets

uint8_t lastMinute = 0;

uint32_t nextSendTime = 0;
uint32_t newRecvTime = 0;
uint32_t lastRecvTime = 0;

uint32_t newTickTime = 0;
uint32_t lastTickTime = 0;

bool rebooted = 1;

uint32_t no_packet_count = 0;


//====================================================================================
//                                    Function prototype
//====================================================================================

void syncTime(void);
void displayTime(void);
void printTime(time_t zone, char *tzCode);
void decodeNTP(void);
void sendNTPpacket(IPAddress& address);

//====================================================================================
//                                    Update Time
//====================================================================================
void syncTime(void)
{
  // Don't send too often so we don't trigger Denial of Service
  if (nextSendTime < millis()) {
    // Get a random server from the pool
    WiFi.hostByName(ntpServerName, timeServerIP);
    nextSendTime = millis() + 5000;

    // Flush old late packets
    while  (udp.parsePacket() > 0)  {                // Is a packet there?
      Serial.println("Reading delayed NTP packet."); // Yes
      udp.read(packetBuffer, NTP_PACKET_SIZE);       // read the packet into the buffer
    }

    sendNTPpacket(timeServerIP); // send an NTP packet to a time server
    decodeNTP();
  }
}

//====================================================================================
// Send an NTP request to the time server at the given address
//====================================================================================
void sendNTPpacket(IPAddress& address)
{
  // Serial.println("sending NTP packet...");
  // set all bytes in the buffer to 0
  memset(packetBuffer, 0, NTP_PACKET_SIZE);
  // Initialize values needed to form NTP request
  // (see URL above for details on the packets)
  packetBuffer[0] = 0b11100011;   // LI, Version, Mode
  packetBuffer[1] = 0;            // Stratum, or type of clock
  packetBuffer[2] = 6;            // Polling Interval
  packetBuffer[3] = 0xEC;         // Peer Clock Precision

  // 8 bytes of zero for Root Delay & Root Dispersion

  packetBuffer[12]  = 49;
  packetBuffer[13]  = 0x4E;
  packetBuffer[14]  = 49;
  packetBuffer[15]  = 52;

  // all NTP fields have been given values, now
  // you can send a packet requesting a timestamp:
  udp.beginPacket(address, 123); //NTP requests are to port 123
  udp.write(packetBuffer, NTP_PACKET_SIZE);
  udp.endPacket();
}

//====================================================================================
// Decode the NTP message and print status to serial port
//====================================================================================
void decodeNTP(void)
{
  timeValid = false;
  uint32_t waitTime = millis() + 500;
  while (millis() < waitTime && !timeValid)
  {
    yield();
    if (udp.parsePacket())
    {
      newRecvTime = millis();

      // We've received a packet, read the data from it
      udp.read(packetBuffer, NTP_PACKET_SIZE); // read the packet into the buffer

      Serial.print("\nNTP response time was : ");
      Serial.print(500 - (waitTime - newRecvTime));
      Serial.println(" ms");

      Serial.print("Time since last sync is: ");
      Serial.print((newRecvTime - lastRecvTime) / 1000.0);
      Serial.println(" s");
      lastRecvTime = newRecvTime;

      // The timestamp starts at byte 40 of the received packet and is four bytes,
      // or two words, long. First, extract the two words:
      unsigned long highWord = word(packetBuffer[40], packetBuffer[41]);
      unsigned long lowWord = word(packetBuffer[42], packetBuffer[43]);

      // Combine the four bytes (two words) into a long integer
      // this is NTP time (seconds since Jan 1 1900):
      unsigned long secsSince1900 = highWord << 16 | lowWord;

      // Now convert NTP Unix time (Seconds since Jan 1 1900) into everyday time:
      // UTC time starts on Jan 1 1970. In seconds the difference is 2208988800:
      utc = secsSince1900 - 2208988800UL;

      setTime(utc);      // Set system clock to utc time (not time zone compensated)

      timeValid = true;

      // Print the hour, minute and second:
      Serial.print("Received NTP UTC time : ");

      uint8_t hh = hour(utc);
      Serial.print(hh); // print the hour (86400 equals secs per day)

      Serial.print(':');
      uint8_t mm = minute(utc);
      if (mm < 10 ) Serial.print('0');
      Serial.print(mm); // print the minute (3600 equals secs per minute)

      Serial.print(':');
      uint8_t ss = second(utc);
      if ( ss < 10 ) Serial.print('0');
      Serial.println(ss); // print the second
    }
  }

  // Keep a count of missing or bad NTP replies

  if ( timeValid ) {
    no_packet_count = 0;
  }
  else
  {
    Serial.println("\nNo NTP reply, trying again in 1 minute...");
    no_packet_count++;
  }

  if (no_packet_count >= 10) {
    no_packet_count = 0;
    // TODO: Flag the lack of sync on the display
    Serial.println("\nNo NTP packet in last 10 minutes");
  }
}

//====================================================================================
//                                  Debug use only
//====================================================================================
void printTime(time_t t, char *tzCode)
{
  String dateString = dayStr(weekday(t));
  dateString += " ";
  dateString += day(t);
  if (day(t) == 1 || day(t) == 21 || day(t) == 31) dateString += "st";
  else if (day(t) == 2 || day(t) == 22) dateString += "nd";
  else if (day(t) == 3 || day(t) == 23) dateString += "rd";
  else dateString += "th";

  dateString += " ";
  dateString += monthStr(month(t));
  dateString += " ";
  dateString += year(t);

  // Print time to serial port
  Serial.print(hour(t));
  Serial.print(":");
  Serial.print(minute(t));
  Serial.print(":");
  Serial.print(second(t));
  Serial.print(" ");
  // Print time t
  Serial.print(tzCode);
  Serial.print(" ");

  // Print date
  Serial.print(day(t));
  Serial.print("/");
  Serial.print(month(t));
  Serial.print("/");
  Serial.print(year(t));
  Serial.print("  ");

  // Now test some other functions that might be useful one day!
  Serial.print(dayStr(weekday(t)));
  Serial.print(" ");
  Serial.print(monthStr(month(t)));
  Serial.print(" ");
  Serial.print(dayShortStr(weekday(t)));
  Serial.print(" ");
  Serial.print(monthShortStr(month(t)));
  Serial.println();
}

//====================================================================================
